addon/-private/system/model/record-data-map.ts

```diff
@@ -0,0 +1,63 @@
+import { DEBUG } from '@glimmer/env';
+
+/**
+ * Generally we use a WeakMap to store the 1:1 mapping between
+ * Record/CreateOptions and RecordData.
+ *
+ * However, we have no guarantee that the CreateOptions we pass to
+ * `factory.create(createOptions)` are the same object that we receive
+ * as the first arg to `Model.init()` (in fact, they are not!).
+ *
+ * For this reason, we use this randomly generated string key to store
+ * the mapping for CreateOptions. This maintains isolation without
+ * resulting in a performance hit.
+ */
+export const RECORD_DATA_KEY = `${Date.now()}-record-data`;
+
+type CreateOptions = object;
+type Record = object;
```

Comment:
There's not much value in creating types like these. Some people have the urge to do this because it makes the code more descriptive, but that's what variable/parameter names are for.

Reply:
It was more to make refactoring easier once we have true interfaces for these things.

Reply:
I would add a `// TODO: create a specific interface` to each site then. In general, if you just create aliases to `object` all over the place, you're introducing unnecessary complexity and the _illusion_ of type-safety.